src/sparse_blas/backends/cusparse/cusparse_handles.hpp

```diff
@@ -0,0 +1,91 @@
+            : detail::generic_dense_vector_handle<cusparseDnVecDescr_t>(cu_descr, value_buffer,
+                                                                        size) {}
+};
+
+struct dense_matrix_handle : public detail::generic_dense_matrix_handle<cusparseDnMatDescr_t> {
+    template <typename T>
+    dense_matrix_handle(cusparseDnMatDescr_t cu_descr, T* value_ptr, std::int64_t num_rows,
+                        std::int64_t num_cols, std::int64_t ld, layout dense_layout)
+            : detail::generic_dense_matrix_handle<cusparseDnMatDescr_t>(
+                  cu_descr, value_ptr, num_rows, num_cols, ld, dense_layout) {}
+
+    template <typename T>
+    dense_matrix_handle(cusparseDnMatDescr_t cu_descr, const sycl::buffer<T, 1> value_buffer,
+                        std::int64_t num_rows, std::int64_t num_cols, std::int64_t ld,
+                        layout dense_layout)
+            : detail::generic_dense_matrix_handle<cusparseDnMatDescr_t>(
+                  cu_descr, value_buffer, num_rows, num_cols, ld, dense_layout) {}
+};
+
+struct matrix_handle : public detail::generic_sparse_handle<cusparseSpMatDescr_t> {
+    template <typename fpType, typename intType>
+    matrix_handle(cusparseSpMatDescr_t cu_descr, intType* row_ptr, intType* col_ptr,
+                  fpType* value_ptr, detail::sparse_format format, std::int64_t num_rows,
+                  std::int64_t num_cols, std::int64_t nnz, oneapi::mkl::index_base index)
+            : detail::generic_sparse_handle<cusparseSpMatDescr_t>(
+                  cu_descr, row_ptr, col_ptr, value_ptr, format, num_rows, num_cols, nnz, index) {}
+
+    template <typename fpType, typename intType>
+    matrix_handle(cusparseSpMatDescr_t cu_descr, const sycl::buffer<intType, 1> row_buffer,
+                  const sycl::buffer<intType, 1> col_buffer,
+                  const sycl::buffer<fpType, 1> value_buffer, detail::sparse_format format,
+                  std::int64_t num_rows, std::int64_t num_cols, std::int64_t nnz,
+                  oneapi::mkl::index_base index)
+            : detail::generic_sparse_handle<cusparseSpMatDescr_t>(cu_descr, row_buffer, col_buffer,
+                                                                  value_buffer, format, num_rows,
+                                                                  num_cols, nnz, index) {}
+};
+
+inline void check_valid_matrix_properties(const std::string& function_name,
+                                          matrix_handle_t sm_handle) {
+    if (sm_handle->format == detail::sparse_format::COO &&
+        !(sm_handle->has_matrix_property(matrix_property::sorted_by_rows) ||
+          sm_handle->has_matrix_property(matrix_property::sorted))) {
+        throw mkl::unimplemented(
+            "sparse_blas", function_name,
+            "The backend does not support unsorted COO format. Use `set_matrix_property` to set the property `matrix_property::sorted_by_rows` or `matrix_property::sorted`");
```

Comment:
I was going to suggest changing `matrix_property` to `property` here, but it looks like we didn't realize when reviewing https://github.com/uxlfoundation/oneAPI-spec/pull/522 that the [oneAPI Specification uses `enum matrix_property`](https://uxlfoundation.github.io/oneAPI-spec/spec/elements/oneMKL/source/domains/spblas/data_types/set_matrix_property.html#onemkl-sparse-set-matrix-property) but the [Intel oneMKL uses `enum property`](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-dpcpp/2024-2/oneapi-mkl-sparse-set-matrix-property.html).

cuSPARSE supports sparse vectors, for which the `sorted` property is applicable. Likewise, (in the far future) the `symmetric` etc. properties are also applicable to dense matrices. (Particularly on the AI side, typical autoregressive self-attention in LLMs leads to dense triangular matrices.)

@spencerpatty, do you think it's worth changing the oneAPI Specifcation to rename `matrix_property` back to `property`, or should we introduce new `vector_property`, `dense_matrix_property` etc. in the future when we eventually get to them?

Reply:
I think it makes sense to use `matrix_property` and `dense_matrix_property` and if we ever get to it, `vector_property` though I doubt that an unsorted sparse vector makes any kind of sense ...

To be complete, I will provide in the next comment the justification for this suggestion as a list of potential properties that sparse matrices and dense matrices might have (and they are quite different sets)

Reply:
Potential properties for sparse matrices might be 

- `sparse::matrix_property::sorted_by_rows` -- partially sorted by rows, but no guarantees within rows
- `sparse::matrix_property::sorted` -- fully sorted as per format natural state, (and sorted_by_rows + sorted within rows for COO)
- `sparse::matrix_property::symmetric` -- full data is provided but it is a symmetric matrix
- `sparse::matrix_property::hermitian` -- full data is provided but it is a hermiatian matrix
- `sparse::matrix_property::lower_triangular` -- L+D lower data is provided (could be used with triangular/symmetric/hermitian matrix view and diag unit/nonunit)
- `sparse::matrix_property::upper_triangular` -- D+U lower data is provided (could be used with triangular/symmetric/hermitian matrix view and diag::unit/nonunit)
- `sparse::matrix_property::strict_lower_triangular` -- strict L lower data is provided (could be used with triangular/symmetric/hermitian matrix view and diag::unit/nonunit)
- `sparse::matrix_property::strict_upper_triangular` -- strict U upper data is provided (could be used with triangular/symmetric/hermitian matrix view and diag::unit/nonunit)
- `sparse::matrix_property::diagonal` -- D diagonal data is provided (could be used with diag::unit/nonunit)

Reply:
Potential properties for dense matrices might incorporate things from BLAS/LAPACK special formats (https://www.netlib.org/lapack/lug/node121.html)

- `sparse::dense_matrix_property::symmetric` -- explicitly filled with full L+D+U data and is symmetric
- `sparse::dense_matrix_property::hermitian` -- explicitly filled with full L+D+U data and is hermitian
- `sparse::dense_matrix_property::lower_triangular` -- explicitly filled with full L+D data using general format with leading dimension (may be used with triangular/symmetric/hermitian views and diag unit/nonunit )
- `sparse::dense_matrix_property::upper_triangular` -- explicitly filled with full D+U data using general format with leading dimension (may be used with triangular/symmetric/hermitian views and diag unit/nonunit )
- `sparse::dense_matrix_property::lower_packed` -- explicitly filled with full L+D data using packed format (may be used with triangular/symmetric/hermitian views and diag unit/nonunit )
- `sparse::dense_matrix_property::upper_packed` -- explicitly filled with full D+U data using packed format (may be used with triangular/symmetric/hermitian views and diag unit/nonunit )
- `sparse::dense_matrix_property::banded` -- filled with data in banded format
- `sparse::dense_matrix_property::lower_banded` -- explicitly filled with lower and main diagonal portion of banded data (may be used with triangular/symmetric/hermitian banded views)
- `sparse::dense_matrix_property::upper_banded` -- explicitly filled with upper and main diagonal portion of banded data (may be used with triangular/symmetric/hermitian banded views)
- `sparse::dense_matrix_property::tridiagonal` -- filled with data in unsymmetric tridiagonal format
- `sparse::dense_matrix_property::symmetric_tridiagonal` -- filled with two 1D arrays one of length n of diagonal elements and one of length n-1 containing off-diagonal elements

though many of these might better be considered as formats for the dense matrix, since we would not get correctness if they were not provided and properties should only be viewed as optimization hints.

So if we take out `general`, `packed`, `banded` and tridiagonal` and `symmetric_tridiagonal` as formats, then we are left with the following potential properties:

- `symmetric` -- data is explicitly symmetric as provided
- `hermitian` -- data is explicitly Hermitian as provided
- `lower` -- data is from L+D in the format provided (can be utilized with triangular/symmetric/hermitian views and diag unit/nonunit)
- `upper` -- data is from D+U in the format provided (can be utilized with triangular/symmetric/hermitian views and diag unit/nonunit)

Does that make sense ?  I have a small internal doubt about lower/upper being properties of the data (and optional) ... maybe someone sees a case where it is not sufficient to be considered a property ?

Reply:
Also note that we wouldn't be able to rename the enum to `property` as it would conflict with the enum defined in the Intel product. We have seen we need to define more properties in the oneMKL Interface enum than the latest release of oneMKL product currently has so they need to have different names.
This is one of the major flaw with trying to have the oneMKL product as both a backend of oneMKL Interface and an implementation of the open-source oneMKL Specification.

Reply:
Okay, let's keep all the properties separate and continue using `matrix_property` on the oneAPI Spec side.